hearth/lib/hearth/middleware/retry.rb

```diff
@@ -2,114 +2,93 @@
+      #     client side throttling. This is a provisional mode that may change
+      #     behavior in the future.
+      def initialize(app, retry_strategy:, error_inspector_class:)
         @app = app
 
-        # public config
-        @retry_mode = retry_mode
-        @max_attempts = max_attempts
-        @adaptive_retry_wait_to_fill = adaptive_retry_wait_to_fill
-
-        # undocumented options
+        @retry_strategy = retry_strategy
+        # undocumented - protocol specific
         @error_inspector_class = error_inspector_class
-        @retry_quota = retry_quota
-        @client_rate_limiter = client_rate_limiter
-
-        # instance state
-        @capacity_amount = nil
-        @retries = 0
       end
 
-      # @param input
-      # @param context
-      # @return [Output]
-      # rubocop:disable Metrics/AbcSize
       def call(input, context)
-        acquire_token
-        output = @app.call(input, context)
-        error_inspector = @error_inspector_class.new(
-          output.error, context.response.status
-        )
-        request_bookkeeping(output, error_inspector)
-
-        return output unless retryable?(context, output, error_inspector)
-
-        return output if @retries >= @max_attempts - 1
-
-        @capacity_amount = @retry_quota.checkout_capacity(error_inspector)
-        return output unless @capacity_amount.positive?
-
-        delay = [Kernel.rand * (2**@retries), MAX_BACKOFF].min
-        Kernel.sleep(delay)
-        retry_request(input, context, output)
+        token = @retry_strategy.acquire_initial_retry_token
+
+        raise 'Unable to make request, too many failures' if token.nil?
+
+        # TODO: determine if loop or nested function call is better here
+        # If the latter, no outer variable is needed.
+        output = nil
+        loop do
+          output = @app.call(input, context)
+
+          if (error = output.error)
+            error_inspector = @error_inspector_class.new(
+              error, context.response
+            )
+            error_info = error_info(error, error_inspector)
+            token = @retry_strategy.refresh_retry_token(token, error_info)
+
+            Kernel.sleep(token.retry_delay) if token
+          else
+            @retry_strategy.record_success(token)
+          end
+          break unless token && output.error
+
+          reset_request(context, output)
+        end
+        output
       end
-      # rubocop:enable Metrics/AbcSize
 
       private
 
-      def acquire_token
-        return if @retry_mode == 'standard'
-
-        # either fail fast or block until a token becomes available
-        # must be configurable
-        # need a maximum rate at which we can send requests (max_send_rate)
-        # is unset until a throttle is seen
-        @client_rate_limiter.token_bucket_acquire(
-          1, wait_to_fill: @adaptive_retry_wait_to_fill
-        )
+      def error_info(error, error_inspector)
```

Comment:
We could skip SRA error_info and instead use error_inspector and pass it into our strategy interface. I originally had it that way, and I think it might work out better, but I've done it this way for review first.

Reply:
Yeah - I think that makes sense to do.  Though I think it makes sense to change the name "error_inspector" to "error_info".  

I think we can then do away with the error_info method and move the error_type and modeled_retryable? onto the errorInfo class (yes, this will be transport specific.  Can be duplicated if needed in the future, or we can create a parent class and inherit... but I think it just makes sense as something an ErrorInfo should know).  

I think we leave the `retryable?` method in the stratagies though and allow them to check the type of error (I don't love that, but does give the strategy room to implement that differently I guess).